File: python-sdks/instrumentations/respan-instrumentation-vertexai/src/respan_instrumentation_vertexai/_otel_emitter.py
"""Emit Vertex AI SDK calls as OTEL ReadableSpan objects."""

from __future__ import annotations

import logging
import time
from typing import Any

from opentelemetry import context as context_api
from opentelemetry import trace
from opentelemetry.semconv._incubating.attributes import (
    gen_ai_attributes as GenAIAttributes,
)
from opentelemetry.semconv_ai import LLMRequestTypeValues, SpanAttributes

from respan_instrumentation_vertexai._constants import (
    ASSISTANT_ROLE,
    CANDIDATES_TOKEN_COUNT_KEY,
    PROMPT_TOKEN_COUNT_KEY,
    ROLE_KEY,
    SYSTEM_INSTRUCTION_KEY,
    TOOLS_KEY,
    TOTAL_TOKEN_COUNT_KEY,
    VERTEXAI_GENERATE_CONTENT_SPAN_NAME,
    VERTEXAI_SYSTEM_NAME,
)
from respan_instrumentation_vertexai._translator import (
    extract_tool_calls,
    extract_tools,
    extract_usage,
    format_input,
    format_output,
    normalize_input_messages,
    safe_json,
    to_json_attr,
)
from respan_sdk.constants.llm_logging import LOG_TYPE_CHAT
from respan_sdk.constants.span_attributes import RESPAN_LOG_TYPE
from respan_sdk.utils.data_processing.id_processing import (
    format_span_id,
    format_trace_id,
)
from respan_tracing.utils.span_factory import build_readable_span, inject_span

logger = logging.getLogger(__name__)


def _current_trace_parent_ids() -> tuple[str | None, str | None]:
    try:
        current_span = trace.get_current_span()
        span_context = current_span.get_span_context()
    except Exception:
        return None, None

    trace_id = getattr(span_context, "trace_id", 0) or 0
    span_id = getattr(span_context, "span_id", 0) or 0
    if trace_id == 0 or span_id == 0:
        return None, None
    return format_trace_id(trace_id=trace_id), format_span_id(span_id=span_id)


def _base_attrs(span_name: str) -> dict[str, Any]:
    attrs = {
        SpanAttributes.LLM_SYSTEM: VERTEXAI_SYSTEM_NAME,
        SpanAttributes.LLM_REQUEST_TYPE: LLMRequestTypeValues.CHAT.value,
        SpanAttributes.TRACELOOP_ENTITY_NAME: span_name,
        SpanAttributes.TRACELOOP_ENTITY_PATH: span_name,
        RESPAN_LOG_TYPE: LOG_TYPE_CHAT,
    }
    workflow_name = context_api.get_value(SpanAttributes.TRACELOOP_ENTITY_NAME)
    if workflow_name:
        attrs[SpanAttributes.TRACELOOP_WORKFLOW_NAME] = workflow_name
    return attrs


def _set_input_attrs(attrs: dict[str, Any], request_payload: dict[str, Any]) -> None:
    contents = request_payload.get("contents")
    system_instruction = request_payload.get(SYSTEM_INSTRUCTION_KEY)
    messages = normalize_input_messages(
        contents=contents,
        system_instruction=system_instruction,
    )
    attrs[SpanAttributes.TRACELOOP_ENTITY_INPUT] = format_input(
        contents=contents,
        system_instruction=system_instruction,
    )
    for index, message in enumerate(messages):
        role = message.get(ROLE_KEY)
        content = message.get("content")
        if role is not None:
            attrs[f"{SpanAttributes.LLM_PROMPTS}.{index}.role"] = str(role)
        if content is not None:
            attrs[f"{SpanAttributes.LLM_PROMPTS}.{index}.content"] = to_json_attr(
                content
            )


def _set_output_attrs(attrs: dict[str, Any], response_or_chunks: Any) -> None:
    output = format_output(response_or_chunks=response_or_chunks)
    attrs[SpanAttributes.TRACELOOP_ENTITY_OUTPUT] = output
    attrs[f"{SpanAttributes.LLM_COMPLETIONS}.0.role"] = ASSISTANT_ROLE
    attrs[f"{SpanAttributes.LLM_COMPLETIONS}.0.content"] = output

    tool_calls = extract_tool_calls(response_or_chunks=response_or_chunks)
    if tool_calls:
        attrs[f"{SpanAttributes.LLM_COMPLETIONS}.0.tool_calls"] = safe_json(
            value=tool_calls
        )

    usage = extract_usage(response_or_chunks=response_or_chunks)
    if PROMPT_TOKEN_COUNT_KEY in usage:
        attrs[GenAIAttributes.GEN_AI_USAGE_INPUT_TOKENS] = usage[PROMPT_TOKEN_COUNT_KEY]
        attrs[SpanAttributes.LLM_USAGE_PROMPT_TOKENS] = usage[PROMPT_TOKEN_COUNT_KEY]
    if CANDIDATES_TOKEN_COUNT_KEY in usage:
        attrs[GenAIAttributes.GEN_AI_USAGE_OUTPUT_TOKENS] = usage[
            CANDIDATES_TOKEN_COUNT_KEY
        ]
        attrs[SpanAttributes.LLM_USAGE_COMPLETION_TOKENS] = usage[
            CANDIDATES_TOKEN_COUNT_KEY
        ]
    if TOTAL_TOKEN_COUNT_KEY in usage:
        attrs[SpanAttributes.LLM_USAGE_TOTAL_TOKENS] = usage[TOTAL_TOKEN_COUNT_KEY]


def _set_request_attrs(attrs: dict[str, Any], request_payload: dict[str, Any]) -> None:
    model = request_payload.get("model")
    if model:
        attrs[SpanAttributes.LLM_REQUEST_MODEL] = model

    tools = extract_tools(request_payload.get(TOOLS_KEY))
    if tools:
        attrs[SpanAttributes.LLM_REQUEST_FUNCTIONS] = safe_json(value=tools)

    _set_input_attrs(attrs=attrs, request_payload=request_payload)


def build_generate_content_attrs(
    *,
    request_payload: dict[str, Any],
    response_or_chunks: Any = None,
    span_name: str = VERTEXAI_GENERATE_CONTENT_SPAN_NAME,
) -> dict[str, Any]:
    attrs = _base_attrs(span_name=span_name)
    _set_request_attrs(attrs=attrs, request_payload=request_payload)
    if response_or_chunks is not None:
        _set_output_attrs(attrs=attrs, response_or_chunks=response_or_chunks)
    return attrs


def emit_generate_content_span(
    *,
    request_payload: dict[str, Any],
    start_ns: int,
    response_or_chunks: Any = None,
    span_name: str = VERTEXAI_GENERATE_CONTENT_SPAN_NAME,
    error_message: str | None = None,
    status_code: int = 200,
) -> None:
    """Build a ReadableSpan for a Vertex AI generation and inject it."""
    try:
        attrs = build_generate_content_attrs(
            request_payload=request_payload,
            response_or_chunks=response_or_chunks,
            span_name=span_name,
        )
        if error_message:
            attrs["error.message"] = error_message
            attrs.setdefault("status_code", status_code if status_code >= 400 else 500)

        trace_id, parent_id = _current_trace_parent_ids()
        span = build_readable_span(
            name=span_name,
            trace_id=trace_id,
            parent_id=parent_id,
            start_time_ns=start_ns,
            end_time_ns=time.time_ns(),
            attributes=attrs,
            error_message=error_message,
            status_code=status_code,
        )
        inject_span(span=span)
    except Exception:
        logger.debug("Failed to emit Vertex AI span", exc_info=True)

File: javascript-sdks/instrumentations/respan-instrumentation-vertexai/src/_otel_emitter.ts
import { context, trace } from "@opentelemetry/api";
import type { ReadableSpan } from "@opentelemetry/sdk-trace-base";
import {
  ATTR_GEN_AI_COMPLETION,
  ATTR_GEN_AI_PROMPT,
  ATTR_GEN_AI_REQUEST_MAX_TOKENS,
  ATTR_GEN_AI_REQUEST_MODEL,
  ATTR_GEN_AI_REQUEST_TEMPERATURE,
  ATTR_GEN_AI_REQUEST_TOP_K,
  ATTR_GEN_AI_REQUEST_TOP_P,
  ATTR_GEN_AI_SYSTEM,
  ATTR_GEN_AI_USAGE_COMPLETION_TOKENS,
  ATTR_GEN_AI_USAGE_INPUT_TOKENS,
  ATTR_GEN_AI_USAGE_OUTPUT_TOKENS,
  ATTR_GEN_AI_USAGE_PROMPT_TOKENS,
} from "@opentelemetry/semantic-conventions/incubating";
import { RespanLogType, RespanSpanAttributes } from "@respan/respan-sdk";
import { buildReadableSpan, getEntityPath, injectSpan, WORKFLOW_NAME_KEY } from "@respan/tracing";
import { SpanAttributes } from "@traceloop/ai-semantic-conventions";
import {
  ASSISTANT_ROLE,
  CANDIDATES_TOKEN_COUNT_KEY,
  GENERATION_CONFIG_KEY,
  INSTRUMENTATION_LIBRARY_NAME,
  PACKAGE_VERSION,
  PROMPT_TOKEN_COUNT_KEY,
  RESPAN_LOG_METHOD_TS_TRACING,
  SYSTEM_INSTRUCTION_KEY,
  TOOLS_KEY,
  TOTAL_TOKEN_COUNT_KEY,
  VERTEXAI_GENERATE_CONTENT_SPAN_NAME,
  VERTEXAI_SYSTEM_NAME,
} from "./_constants.js";
import {
  extractToolCalls,
  extractTools,
  extractUsage,
  formatInput,
  formatOutput,
  normalizeInputMessages,
  safeJson,
  toJsonAttr,
  type VertexAIRequestPayload,
} from "./_translator.js";

const GEN_AI_COMPLETION_PREFIX = `${ATTR_GEN_AI_COMPLETION}.0`;
const LLM_USAGE_TOTAL_TOKENS = SpanAttributes.LLM_USAGE_TOTAL_TOKENS;

function maybeSetAttr(attrs: Record<string, unknown>, key: string, value: unknown): void {
  if (value === undefined || value === null || value === "") return;
  if (Array.isArray(value) && value.length === 0) return;
  attrs[key] = value as any;
}

function numberField(value: unknown, ...names: string[]): number | undefined {
  if (!value || typeof value !== "object") return undefined;
  for (const name of names) {
    const fieldValue = (value as Record<string, unknown>)[name];
    if (typeof fieldValue === "number") return fieldValue;
  }
  return undefined;
}

function generationConfigValue(payload: VertexAIRequestPayload, ...names: string[]): number | undefined {
  return numberField(payload.generationConfig, ...names);
}

function buildInstrumentedReadableSpan(opts: {
  name: string;
  startTimeIso: string;
  attributes: Record<string, unknown>;
  errorMessage?: string;
  statusCode?: number;
}): ReadableSpan {
  const activeSpanContext = trace.getSpan(context.active())?.spanContext();
  const span = buildReadableSpan({
    name: opts.name,
    traceId: activeSpanContext?.traceId,
    parentId: activeSpanContext?.spanId,
    startTimeIso: opts.startTimeIso,
    endTimeIso: new Date().toISOString(),
    attributes: opts.attributes,
    errorMessage: opts.errorMessage,
    statusCode: opts.statusCode,
  }) as ReadableSpan & {
    instrumentationLibrary?: { name: string; version?: string };
  };

  span.instrumentationLibrary = {
    name: INSTRUMENTATION_LIBRARY_NAME,
    version: PACKAGE_VERSION,
  };
  return span;
}

function baseAttrs(spanName: string): Record<string, unknown> {
  const activeContext = context.active();
  const workflowName = activeContext.getValue(WORKFLOW_NAME_KEY) as string | undefined;
  const entityPath = getEntityPath(activeContext);
  const attrs: Record<string, unknown> = {
    [SpanAttributes.TRACELOOP_ENTITY_NAME]: spanName,
    [SpanAttributes.TRACELOOP_ENTITY_PATH]: entityPath || spanName,
    [RespanSpanAttributes.RESPAN_LOG_METHOD]: RESPAN_LOG_METHOD_TS_TRACING,
    [RespanSpanAttributes.RESPAN_LOG_TYPE]: RespanLogType.CHAT,
    [SpanAttributes.LLM_REQUEST_TYPE]: RespanLogType.CHAT,
    [ATTR_GEN_AI_SYSTEM]: VERTEXAI_SYSTEM_NAME,
  };

  if (workflowName) {
    attrs[SpanAttributes.TRACELOOP_WORKFLOW_NAME] = workflowName;
  }
  return attrs;
}

function setInputAttrs(attrs: Record<string, unknown>, requestPayload: VertexAIRequestPayload): void {
  const messages = normalizeInputMessages(
    requestPayload.contents,
    requestPayload.systemInstruction,
  );
  attrs[SpanAttributes.TRACELOOP_ENTITY_INPUT] = formatInput(
    requestPayload.contents,
    requestPayload.systemInstruction,
  );

  messages.forEach((message, index) => {
    const role = message.role;
    const content = message.content;
    maybeSetAttr(attrs, `${ATTR_GEN_AI_PROMPT}.${index}.role`, role);
    if (content !== undefined) {
      attrs[`${ATTR_GEN_AI_PROMPT}.${index}.content`] = toJsonAttr(content);
    }
  });
}

function setRequestAttrs(attrs: Record<string, unknown>, requestPayload: VertexAIRequestPayload): void {
  maybeSetAttr(attrs, ATTR_GEN_AI_REQUEST_MODEL, requestPayload.model);
  maybeSetAttr(
    attrs,
    ATTR_GEN_AI_REQUEST_MAX_TOKENS,
    generationConfigValue(requestPayload, "maxOutputTokens", "max_output_tokens"),
  );
  maybeSetAttr(
    attrs,
    ATTR_GEN_AI_REQUEST_TEMPERATURE,
    generationConfigValue(requestPayload, "temperature"),
  );
  maybeSetAttr(
    attrs,
    ATTR_GEN_AI_REQUEST_TOP_P,
    generationConfigValue(requestPayload, "topP", "top_p"),
  );
  maybeSetAttr(
    attrs,
    ATTR_GEN_AI_REQUEST_TOP_K,
    generationConfigValue(requestPayload, "topK", "top_k"),
  );

  const tools = extractTools(requestPayload.tools);
  if (tools.length > 0) {
    attrs[SpanAttributes.LLM_REQUEST_FUNCTIONS] = safeJson(tools);
  }

  setInputAttrs(attrs, requestPayload);

  if (requestPayload.generationConfig !== undefined) {
    attrs[`vertexai.${GENERATION_CONFIG_KEY}`] = safeJson(requestPayload.generationConfig);
  }
  if (requestPayload.toolConfig !== undefined) {
    attrs["vertexai.toolConfig"] = safeJson(requestPayload.toolConfig);
  }
  if (requestPayload.systemInstruction !== undefined) {
    attrs[`vertexai.${SYSTEM_INSTRUCTION_KEY}`] = safeJson(requestPayload.systemInstruction);
  }
  if (requestPayload.tools !== undefined) {
    attrs[`vertexai.${TOOLS_KEY}`] = safeJson(requestPayload.tools);
  }
}

function setOutputAttrs(attrs: Record<string, unknown>, responseOrChunks: unknown): void {
  const output = formatOutput(responseOrChunks);
  attrs[SpanAttributes.TRACELOOP_ENTITY_OUTPUT] = output;
  attrs[`${GEN_AI_COMPLETION_PREFIX}.role`] = ASSISTANT_ROLE;
  attrs[`${GEN_AI_COMPLETION_PREFIX}.content`] = output;

  const toolCalls = extractToolCalls(responseOrChunks);
  if (toolCalls.length > 0) {
    attrs[`${GEN_AI_COMPLETION_PREFIX}.tool_calls`] = safeJson(toolCalls);
  }

  const usage = extractUsage(responseOrChunks);
  const promptTokens = usage[PROMPT_TOKEN_COUNT_KEY];
  const completionTokens = usage[CANDIDATES_TOKEN_COUNT_KEY];
  const totalTokens = usage[TOTAL_TOKEN_COUNT_KEY];

  if (promptTokens !== undefined) {
    attrs[ATTR_GEN_AI_USAGE_INPUT_TOKENS] = promptTokens;
    attrs[ATTR_GEN_AI_USAGE_PROMPT_TOKENS] = promptTokens;
  }
  if (completionTokens !== undefined) {
    attrs[ATTR_GEN_AI_USAGE_OUTPUT_TOKENS] = completionTokens;
    attrs[ATTR_GEN_AI_USAGE_COMPLETION_TOKENS] = completionTokens;
  }
  if (totalTokens !== undefined) {
    attrs[LLM_USAGE_TOTAL_TOKENS] = totalTokens;
  }
}

export function buildGenerateContentAttrs(opts: {
  requestPayload: VertexAIRequestPayload;
  responseOrChunks?: unknown;
  spanName?: string;
}): Record<string, unknown> {
  const attrs = baseAttrs(opts.spanName ?? VERTEXAI_GENERATE_CONTENT_SPAN_NAME);
  setRequestAttrs(attrs, opts.requestPayload);
  if (opts.responseOrChunks !== undefined) {
    setOutputAttrs(attrs, opts.responseOrChunks);
  }
  return attrs;
}

export function emitGenerateContentSpan(opts: {
  requestPayload: VertexAIRequestPayload;
  startTimeIso: string;
  responseOrChunks?: unknown;
  spanName?: string;
  errorMessage?: string;
  statusCode?: number;
}): void {
  try {
    const spanName = opts.spanName ?? VERTEXAI_GENERATE_CONTENT_SPAN_NAME;
    const attrs = buildGenerateContentAttrs({
      requestPayload: opts.requestPayload,
      responseOrChunks: opts.responseOrChunks,
      spanName,
    });
    if (opts.errorMessage) {
      attrs["error.message"] = opts.errorMessage;
      attrs["status_code"] = opts.statusCode ?? 500;
    }

    injectSpan(buildInstrumentedReadableSpan({
      name: spanName,
      startTimeIso: opts.startTimeIso,
      attributes: attrs,
      errorMessage: opts.errorMessage,
      statusCode: opts.statusCode,
    }));
  } catch {
    // Instrumentation must never break application code.
  }
}
